src/app/core/products.service.ts

```diff
@@ -2,23 +2,44 @@ import { Injectable } from '@angular/core';
 import { HttpClient } from '@angular/common/http';
 import { Observable, of } from 'rxjs';
 import { catchError } from 'rxjs/operators';
-import {Product} from '../product.interface';
+import { Product } from '../product.interface';
 
 @Injectable({
-  providedIn: 'root'
+  providedIn: 'root',
 })
 export class ProductsService {
+  private readonly getProductsUrl = '/api/products';
+  isLoading: boolean = false;
 
-  private readonly getProductsUrl = '/products';
-
-  constructor(private _http: HttpClient) { }
+  constructor(private http: HttpClient) {}
 
   getAllProducts(): Observable<Product[]> {
-    return this._http
+    return this.http
       .get<Product[]>(this.getProductsUrl)
       .pipe(catchError(this.handleError<Product[]>([])));
   }
-  
+
+  deleteProduct(id: number): Observable<any> {
+    this.isLoading = true;
+    return this.http
+      .delete(`${this.getProductsUrl}/${id}`)
+      .pipe(catchError(this.handleError()));
+  }
+
+  addProduct(productData: Product): Observable<Product> {
+    this.isLoading = true;
+    return this.http
+      .post<Product>(this.getProductsUrl, productData)
+      .pipe(catchError(this.handleError<any>({})));
+  }
+
+  updateProduct(id: number, productData: Product): Observable<Product> {
+    this.isLoading = true;
+    return this.http
+      .put<Product>(`${this.getProductsUrl}/${id}`, productData)
+      .pipe(catchError(this.handleError<any>({})));
```

Comment:
Consider if such error handling is useful. Errors are already logged into the console. And `of` in `handleError` silences any errors to be handled. (components will not be able to handle errors)